File: clients/python/pocketsql_client.py
import requests
import json
from tabulate import tabulate

BASE_URL = "http://<mobile_ip>:<active_port>/api/query"
API_KEY = "<your_api_key>"
DATABASE = "ecommerce"

HEADERS = {
    "Authorization": f"Bearer {API_KEY}",
    "Content-Type": "application/json"
}

def run_sql(sql):
    sql = sql.strip()
    if not sql:
        return

    print(f"\nmysql> {sql}\n")

    try:
        payload = {"sql": sql, "database": DATABASE}
        response = requests.post(BASE_URL, json=payload, headers=HEADERS, timeout=10)

        try:
            response.raise_for_status()
            data = response.json()
        except requests.exceptions.RequestException as e:
            print("ERROR:", str(e))
            return
        except json.JSONDecodeError:
            print("ERROR: Invalid JSON from server")
            if response.text:
                print(response.text)
            return

        if data.get("success"):
            columns = data.get("columns", [])
            rows = data.get("rows", [])
            exec_time = data.get("executionTimeMs", 0) / 1000.0

            if rows:
                if isinstance(rows[0], dict):
                    print(tabulate(rows, headers="keys", tablefmt="psql"))
                else:
                    print(tabulate(rows, headers=columns, tablefmt="psql"))
                print(f"\n{len(rows)} rows in set ({exec_time:.2f} sec)")
            else:
                print("Query OK")
                if "affected_rows" in data:
                    print(f"{data['affected_rows']} rows affected")
        else:
            print("ERROR:")
            print(data.get("error") or data.get("message") or json.dumps(data))

    except Exception as e:
        print("ERROR:", str(e))

print("=" * 50)
print("        PocketSQL Terminal (Python)")
print("=" * 50)
print("Type 'exit' to quit\n")

while True:
    sql = input("mysql> ")
    if sql.lower() in ["exit", "quit"]:
        print("\nBye")
        break
    run_sql(sql)

File: clients/javascript/pocketsql_client.js
const BASE_URL = 'http://<mobile_ip>:<active_port>/api/query';
const API_KEY = '<your_api_key>';
const DATABASE = 'ecommerce';

const readline = require('readline');

function runSql(sql) {
  sql = sql.trim();
  if (!sql) return;

  console.log(`\nmysql> ${sql}\n`);

  const payload = JSON.stringify({ sql, database: DATABASE });

  fetch(BASE_URL, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${API_KEY}`,
      'Content-Type': 'application/json'
    },
    body: payload
  })
  .then(async res => {
    const body = await res.text();
    if (!res.ok) {
      console.log(`ERROR: HTTP ${res.status}\n${body}`);
      return;
    }

    let data;
    try { data = JSON.parse(body); } catch {
      console.log('ERROR: Invalid JSON from server\n' + body);
      return;
    }

    if (!data.success) {
      console.log('ERROR: ' + (data.message || data.error || body));
      return;
    }

    const columns = data.columns || [];
    const rows = data.rows || [];
    const execTime = (data.executionTimeMs || 0) / 1000;

    if (!rows.length) {
      console.log('Query OK');
      if (data.affected_rows) console.log(data.affected_rows + ' rows affected');
      return;
    }

    const isDict = typeof rows[0] === 'object' && !Array.isArray(rows[0]);
    let widths = columns.map(c => c.length);
    const tableRows = [];

    for (const row of rows) {
      const r = [];
      for (const col of columns) {
        r.push(isDict ? (row[col] ?? '') : String(row.shift ? row.shift() : ''));
      }
      tableRows.push(r);
      for (let i = 0; i < r.length; i++) {
        if (r[i].length > widths[i]) widths[i] = r[i].length;
      }
    }

    const sep = '+' + widths.map(w => '-'.repeat(w + 2)).join('+') + '+';

    console.log(sep);
    console.log('|' + columns.map((c, i) => ` ${c.padEnd(widths[i])} |`).join(''));
    console.log(sep);

    for (const row of tableRows) {
      console.log('|' + row.map((v, i) => ` ${v.padEnd(widths[i])} |`).join(''));
    }
    console.log(sep);
    console.log(`\n${rows.length} rows in set (${execTime.toFixed(2)} sec)`);
  })
  .catch(err => console.log('ERROR: ' + err.message));
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

console.log('='.repeat(50));
console.log('        PocketSQL Terminal (JS)');
console.log('='.repeat(50));
console.log("Type 'exit' to quit\n");

rl.setPrompt('mysql> ');
rl.prompt();

rl.on('line', line => {
  const sql = line.trim();
  if (sql.toLowerCase() === 'exit' || sql.toLowerCase() === 'quit') {
    console.log('\nBye');
    rl.close();
    return;
  }
  runSql(sql);
  setTimeout(() => rl.prompt(), 100);
});
